### ISSUE TITLE
[Feature Request]: GMP Module mpf_t string support

### ISSUE THREAD

**Issue by s-timcheck** [chapel-lang/chapel#25082]
### Summary of Feature

**Description:**

The GMP Module does not have support for getting strings of the `mpf_t` type, unlike the facilities it has for `mpz_t`.  There exists `chpl_gmp_mpz_get_str` which uses `mpz_get_str`, but neither `chpl_gmp_mpf_get_str` or `mpf_get_str` are implemented in the current version of Chapel, so there is no way to get a `string` conversion for the `mpf_t` type.  This will probably require adding `mp_exp_t` as well?

**Is this a blocking issue with no known work-arounds?**

The only workaround I see for printing `mpf_t` vars is redefining `gmp_printf` where the `args` parameter has the designated intent of `const ref`:

`extern proc gmp_printf(fmt: c_ptrConst(c_char), const ref arg ...);`

The designated intent is probably an issue for the `mpz_t` type too.

### Code Sample
I would like to be able to do something like below:
```chapel
var x:mpf_t;
mpf_init(x);
mpf_set_d(x, 0.123456789);

var s:c_ptrConst(c_char) = mpf_get_str(s, 1, 10, 4, x); //string s, exponent of 1, base 10, print 4 digits, mpf_t operand
```

**Comment by bradcray** [comment]
Hi @s-timcheck — As you're finding, we have primarily used the GMP module to implement Chapel's `bigint` type, such that other aspects of GMP have received little, if any, attention.  That said, there's nothing preventing us from building out more of it apart from (up until now, maybe?) lack of historical interest from users.  So if you'd like to see this support built out more completely, it might be useful to spawn off additional issues about that.

For the example you posted in your OP, I took a quick swing at how quickly I could get that up and running, with the caveat that I've never used these routines nor the `mpf_t` type.  I came up with this [[ATO](https://ato.pxeger.com/run?1=fVHNSsQwEGavfYphQWghll131bXFkwdPggePQqjdaQ2sacjMSvQhfAEve9GH8mmcbFvEBQ0JSWa--b75ef-sHyuHm93uY8vN8epr8rYlhKu7F4ek4PrmtkwSDIzeAosNnpzG4DTDJRjL2PpqU44A57taAI1ukTWxT-UUUGsnz1qLkM8U1J0lBo8NCI-LgJFSJfDHeqgII5EoKrB6bVrDFA1kXvG_wF7MWOhcsc-Ms4OEfsrbZ--8aDRp5VvK8zx6nysPYQguk3gZazgNWf8hKXWdBgWzfH6yWJ6ena8uxjAqDqRIuvarPwrmsmcKlgoi4yA_PaJ7O1VAWdmPZZjOOKVv)]:

```chapel
use CTypes, GMP;

extern type mp_exp_t = integral;
extern proc mpf_get_str(str: c_ptr(c_char), const ref expptr: mp_exp_t,
                        base: c_int, n_digits: c_size_t,
                        const in op: mpf_t): c_ptr(c_char);

extern proc printf(args...);

var x: mpf_t;
mpf_init(x);
mpf_set_d(x, 0.123456789);

var s:c_ptr(c_char);
s = mpf_get_str(s, 1, 10, 4, x);
printf("%s\n", s);
```

which gives me this:

```console
1235
```

which seems as though it could conceivably be right?  (I started to read about the implied placement of the decimal point, then got called into a meeting, then decided you'd know immediately what was right or wrong rather than going back to the manual).

I actually wrote up a few versions of the routine, as I'm always wrestling with whether to do "pure" transliterations of C routines (like using `c_ptr(mp_exp_t)` as a type) or to make them more Chapeltasitic (e.g., using a `ref` or `const ref` argument instead) and in this case, I came up with something of a mix, mostly due to wanting the input/output to be `c_ptr(c_char)`.  If you're interested in seeing or considering the other variations, I can share those as well.

**Comment by s-timcheck** [comment]
Hi @bradcray, thank you for getting back on this issue!  Yes, it would be nice to have full `mpf_t` support, but for now everything else I require appears to be currently supported.

Yes, that appears to be the correct behavior.  I am not a GMP expert either, but running some quick tests in C, `mpf_get_str` returns the floating point number as a string without the radix point, and returns the radix point's location in the `exponent` field.  It is the work of the user to format the string with the radix point.

In the example I sent, I had the `exponent` set to `1` so the radix point location in the string isn't getting returned to anywhere proper.  I didn't know how implementing `mp_exp_t` would work in Chapel, so I just left that field as `1`.  When using `mpf_get_str`, the resulting string is rounded up, so printing `1235` is the correct behavior in this case.  If one were to have an `mp_exp_t` var in the `exponent` parameter, it would contain `0` after running in this case.

I'd be curious to see what other routine ideas you came up with?  As long as the exponent can be stored in an `mp_exp_t` type, or even an `int`, you can construct the floating point output.

**Comment by bradcray** [comment]
I've put a draft PR together [here](https://github.com/chapel-lang/chapel/pull/25124) and would be curious if you have any comments on the test cases I added, @s-timcheck (see mpf_get_str.chpl and .good under the 'Files' tab).

I need to do a bit more testing on this before being confident in it.  Specifically, in my first draft above, I'd missed that the 'expptr' argument serves as a form of output, and am not sure the way I've done it is both portable and safe.

**Comment by s-timcheck** [comment]
Hi @bradcray, no worries!  You are right, it is probably better to leave the `mp_exp_t` type as effectively a `c_long`, since I haven't seen it as anything other than a placeholder for `int` or `long` in all the cases I've seen before.  My guess is that `mp_exp_t` exists in case they ever needed functionality to get really large radix point location values.  Your test cases all look good to me too!  That is interesting that allocating `mp_exp_t` as `int(32)` causes it to not function properly though.  Thank you for your changes!  Don't know if I should close this or if you would like to do more testing first?

**Comment by bradcray** [comment]
Thanks for taking a look, @s-timcheck.  This issue should automatically close when we merge #25124, so no need for you to do anything here.

**Comment by bradcray** [comment]
@s-timcheck : Both of the changes you requested to the GMP module have now been merged, should work on `main`, and be included in our June 2.1 release.

If you're not tracking the main development branch (and don't want to), I believe you should also be able to grab a copy of [GMP.chpl](https://github.com/chapel-lang/chapel/blob/main/modules/standard/GMP.chpl) and either replace your local copy, or direct Chapel to use it from a different directory.  Let me know if you want to try this and need more tips on how to do so.

By all means, let us know if you find any additional problems or concerns with these changes or the GMP module more generally—thanks!

**Comment by s-timcheck** [comment]
@bradcray Great, thank you for the changes!  I'll grab a new copy of GMP.chpl in the meantime and then upgrade to 2.1 once it is released.  Thank you again for your assistance!
